src/Test.UnitTests.Sarif/EnumeratedArtifactTests.cs

```diff
@@ -34,6 +34,43 @@ public void EnumeratedArtifact_BinaryFile_OnDisk()
             ValidateBinaryArtifact(artifact, fileSize);
         }
 
+        [Fact]
+        public void EnumeratedArtifact_FilePath_IllegalChars()
+        {
+            string originalString = "C:\\test" + Path.GetInvalidPathChars()[0] + "ABC.md";
+            var uri = new Uri(originalString, UriKind.Absolute);
+
+            var artifact = new EnumeratedArtifact(FileSystem.Instance) { Uri = uri };
+
+            artifact.CleanPath.Should().Be(originalString);
+        }
+
+
+        [Fact]
+        public void EnumeratedArtifact_FilePath_EncodedChars()
+        {
+            string originalString = "C:\\Users\\%28test%29\\Downloads\\New%2DYear.md";
+            var uri = new Uri(originalString, UriKind.Absolute);
```

Comment:
I see the culprit is the Uri constructor. :(

I'm paging stuff back in here. I tried discussing it with Copilot and after 6 of its suggestions failed, it started to suggest making a `CustomUri` class. 🤣

Reply:
I will contact my dear old friends on the BCL team about this.

Reply:
They were still asleep and I was sniped. Here is a helper I just wrote to make a correct URI from a local absolute path. 

[Demo](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMA6AEQEsBDAcygHsBnYKgY2YG49CTyAkgHkeuXun4B2PAG88RBfwBsJACxEAsjSpQAFAEp5iubkVn+xAA41gACyIBeEqgBEAYRAAdYAFNWngDkfAHcAUnQKAE0fGjAXUXNFADdYogAjGjgAVTAqRyIoEKIcql1rOxhi3IBpHTgyAEE05kYAGwhffQTEohSwIjpGRmzc/JKAMUYwJpb23wAFG1sypa68I0TSAE5dABIXctsQIhlDgF8XNdMe7YNuzcwd/YyRqmOZF5KLq57+J5dBsMSu9Aa9vvdzLcfjdHnsXJ9cmQADKMdg0VqLOzvBFUZGo9GY2zgjaQ2H7UElPFojFLEFDV5UglLYnXBRndasogAeh5RAAPMwIABbIWxACeAD4STyuUQ3GAYr4iDRigAlAREABmYEYQuVUGVzTaHR8WqorVNhzIRGlvKyzFNdiozCIOlYMTgREYmv5DtN7AVmocLkpJV0pAIlRKtSgcH0Li5EqIwEYRFsNFjFqIAHJQtnXQbNebLUsXewpgr2MBWmKyLbZXyuYKReKpZzSCoSlUqJNpka5j5CeHMFZViSTL8qD6VnZHE4oBBWq1DJyzBPfoo7DrggUig0wHRhT4oMAAovWgBRAAe7B8ljYjD0UBoQp83pntn00MS7LEq8UU5ELoACEhJkAIzCEqqQy+HAH5fiSa6IT0W6MDuhQ7vuh6vie163veVCProLiEkQQoQKw6SmjQ/YmmQLiVM+r7vocX4QmYv7IQovIBFMoqtFQABeJazhmnoUaampTMEsSeswrQ0Mwth+FxRCHPkYEAOI+MA4znkOrFkKqd4KbeujZp4njZpU2ZctmVyqbyDRwJ6FqZDodBEPJin2CmRAAOp1GhLo6owwDWj4ZB0Na7ggFyvisC4RAALRJgmHjxX4wAuKpgEgVaADKwCxMAzCBXYxFcpcK6/OuG5qUs+QJkliANXY7GKJx/7cbKWSqkiyXHuWcA+JUaQdEQADWPh3l5ClKX4+qeuWrSPswqnqU4lIXswaKWD4FA2DQRW5FAdDwUZJk0GZLjhOMDFEM1+iXZYpk+MRoQAMwNA9LggJcoiObKpHOgUaHKr06JUJ6fUaup5ZCpYj7HsAS1EME/oZukEDmp6KpdowaQAFY+FWWo6nqVDhapqCSLuO4lAAQjjrQjf0MhEAV7DKa+TVFhaD0ABIsKjTguA9pFOG19hnGQJQQr+XV4EAA)

```cs
 /// <summary>
 /// Create a URI from an absolute file path. 
 /// Use this instead of <see cref="Uri.Uri(string, UriKind)"/> to handle '%' in file paths correctly.
 /// </summary>
 static Uri UriForAbsolutePath(string path)
 {
     if (path == null)
     {
         throw new ArgumentNullException(nameof(path));
     }

     if (!Path.IsPathRooted(path))
     {
         throw new ArgumentException("Path must be absolute.", nameof(path));
     }

     // Normalize path and use forward slashes
     path = Path.GetFullPath(path).Replace('\\', '/');

     // Add leading slash to Windows root. e.g. "C:/test" -> "/C:/test"
     if (!path.StartsWith("/"))
     {
         path = "/" + path;
     }

     // URL-encode, but keep slashes and colons
     path = Uri.EscapeDataString(path).Replace("%2F", "/").Replace("%3A", ":");

     // Path is now a valid URI path component and we can build a Uri object from it.
     return new UriBuilder { Scheme = "file", Host = "", Path =  path }.Uri;
 }
 ```

Reply:
We should use this wherever we construct these URIs from paths and then we can safely use Uri.LocalPath.

Reply:
Great! Can you please just create a clean PR with your work and tests?

Reply:
Ling will be fine with this, I am sure. This PR is months old and she is busy with other things. I ran into this issue again doing some explorations around zip scanning (where we encounter unusual characters). I remembered the open PR, and bumped it. Looking it the implementation, however, I was not certain we can accept it as is. Sounds like you could help move this over the finish line. I'm closing this PR for  now.

Reply:
Will do!